vyper/context/types/bases.py

```diff
@@ -199,6 +200,13 @@ def __init__(
         self.is_immutable = is_immutable
         self.is_public = is_public
 
+    @property
+    def canonical_type(self) -> str:
```

Comment:
So, this is what will generate ABI types in type strings? (e.g. `vyper/codegen/abi.py`)

Reply:
Indeed.  And needed to form function signatures.